src/event-engine/dispatcher.ts

```diff
@@ -70,77 +77,87 @@ export class EventEngineDispatcher extends Dispatcher<effects.Effects, Dependenc
+    this.on(
+      effects.reconnect.type,
+      asyncHandler(async (payload, abortSignal, { receiveEvents, delay, config }) => {
+        if (config.retryConfiguration && config.retryConfiguration.shouldRetry(payload.reason, payload.attempts)) {
+          abortSignal.throwIfAborted();
+
+          await delay(config.retryConfiguration.getDelay(payload.attempts));
+
+          abortSignal.throwIfAborted();
+
+          try {
+            const result = await receiveEvents({
+              abortSignal: abortSignal,
+              channels: payload.channels,
+              channelGroups: payload.groups,
+              timetoken: payload.cursor.timetoken,
+              region: payload.cursor.region,
+              filterExpression: config.filterExpression,
+            });
+
+            return engine.transition(events.reconnectingSuccess(result.metadata, result.messages));
+          } catch (error) {
+            if (error instanceof Error && error.message === 'Aborted') {
+              return;
+            }
+
+            if (error instanceof PubNubError) {
+              return engine.transition(events.reconnectingFailure(error));
+            }
           }
+        } else {
+          return engine.transition(events.reconnectingGiveup());
         }
       }),
     );
 
     this.on(
       effects.handshakeReconnect.type,
-      asyncHandler(async (payload, abortSignal, { handshake, shouldRetry, getRetryDelay, delay }) => {
-        if (!shouldRetry(payload.reason, payload.attempts)) {
-          return engine.transition(events.handshakingReconnectingGiveup());
-        }
-
-        abortSignal.throwIfAborted();
-
-        await delay(getRetryDelay(payload.attempts));
-
-        abortSignal.throwIfAborted();
-
-        try {
-          const result = await handshake({
-            abortSignal: abortSignal,
-            channels: payload.channels,
-            channelGroups: payload.groups,
-          });
-
-          return engine.transition(events.handshakingReconnectingSuccess(result.metadata));
-        } catch (error) {
-          if (error instanceof Error && error.message === 'Aborted') {
-            return;
-          }
-
-          if (error instanceof PubNubError) {
-            return engine.transition(events.handshakingReconnectingFailure(error));
+      asyncHandler(async (payload, abortSignal, { handshake, delay, presenceState, config }) => {
+        if (config.retryConfiguration && config.retryConfiguration.shouldRetry(payload.reason, payload.attempts)) {
+          abortSignal.throwIfAborted();
+
+          await delay(config.retryConfiguration.getDelay(payload.attempts));
+
+          abortSignal.throwIfAborted();
+
+          try {
+            const result = await handshake({
+              abortSignal: abortSignal,
+              channels: payload.channels,
+              channelGroups: payload.groups,
+              filterExpression: config.filterExpression,
+              state: presenceState,
```

Comment:
Can we use the `config.maintainPresenceState` flag and don't pass `presenceState` into `handshake`?

Reply:
it's conditional now ✔️ 
same as with heartbeat

https://github.com/pubnub/javascript/blob/c35559fef8bfb49b63d334fd42fb3733f36c75bd/src/event-engine/dispatcher.ts#L37